x-pack/plugins/security/server/routes/api/public/roles/put.js

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { pick, identity } from 'lodash';
+import Joi from 'joi';
+import { ALL_RESOURCE } from '../../../../../common/constants';
+import { wrapError } from '../../../../lib/errors';
+
+export const schema = Joi.object().keys({
+  metadata: Joi.object().optional(),
+  elasticsearch: Joi.object().keys({
+    cluster: Joi.array().items(Joi.string()),
+    indices: Joi.array().items({
+      names: Joi.array().items(Joi.string()),
+      field_security: Joi.object().keys({
+        grant: Joi.array().items(Joi.string()),
+        except: Joi.array().items(Joi.string()),
+      }),
+      privileges: Joi.array().items(Joi.string()),
+      query: Joi.string().allow(''),
+    }),
+    run_as: Joi.array().items(Joi.string()),
+  }),
+  kibana: Joi.array().items({
+    privileges: Joi.array().items(Joi.string()),
+  }),
+});
+
+const transformKibanaPrivilegeToEs = (application, kibanaPrivilege) => {
+  return {
+    privileges: kibanaPrivilege.privileges,
+    application,
+    resources: [ALL_RESOURCE],
+  };
+};
+
+const transformRolesToEs = (
+  application,
+  payload,
+  existingApplications = []
+) => {
+  const { elasticsearch = {}, kibana = [] } = payload;
+  const otherApplications = existingApplications.filter(
+    roleApplication => roleApplication.application !== application
+  );
+
+  return pick({
+    metadata: payload.metadata,
+    transient_metadata: payload.transient_metadata,
+    cluster: elasticsearch.cluster || [],
+    indices: elasticsearch.indices || [],
+    run_as: elasticsearch.run_as || [],
+    applications: [
+      ...kibana.map(kibanaPrivilege =>
+        transformKibanaPrivilegeToEs(application, kibanaPrivilege)
+      ),
+      ...otherApplications,
+    ],
+  }, identity);
+};
+
+export function initPutRolesApi(
+  server,
+  callWithRequest,
+  routePreCheckLicenseFn,
+  application
+) {
+  server.route({
+    method: 'PUT',
+    path: '/api/security/role/{name}',
+    async handler(request, reply) {
+      const name = request.params.name;
+      try {
+        const existingRoleResponse = await callWithRequest(request, 'shield.getRole', {
+          name,
+          ignore: [404],
+        });
+
+        const body = transformRolesToEs(
```

Comment:
We can add validation for this, sure.

I'm not too worried about someone maliciously doing this, as they'd have to have the cluster `manage_security` privilege and have future knowledge about versions of Kibana, so it's more to prevent an accidental privilege that isn't actually used.